front50-core/src/main/groovy/com/netflix/spinnaker/front50/listeners/ChaosMonkeyEventListener.java

```diff
@@ -0,0 +1,178 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.front50.listeners;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.netflix.spinnaker.fiat.model.Authorization;
+import com.netflix.spinnaker.fiat.model.resources.Permissions;
+import com.netflix.spinnaker.front50.config.ChaosMonkeyEventListenerConfigurationProperties;
+import com.netflix.spinnaker.front50.events.ApplicationEventListener;
+import com.netflix.spinnaker.front50.events.ApplicationPermissionEventListener;
+import com.netflix.spinnaker.front50.model.application.Application;
+import com.netflix.spinnaker.front50.model.application.ApplicationDAO;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+/**
+ * Ensures that when Chaos Monkey is enabled (or disabled) on an Application, its permissions are
+ * applied correctly.
+ *
+ * <p>This listens on both Application events, as well as ApplicationPermission events.
+ */
+@Component
+public class ChaosMonkeyEventListener
+    implements ApplicationEventListener, ApplicationPermissionEventListener {
+  private static final Logger log = LoggerFactory.getLogger(ChaosMonkeyEventListener.class);
+
+  private final ApplicationDAO applicationDAO;
+  private final ChaosMonkeyEventListenerConfigurationProperties properties;
+  private final ObjectMapper objectMapper;
+
+  public ChaosMonkeyEventListener(
+      ApplicationDAO applicationDAO,
+      ChaosMonkeyEventListenerConfigurationProperties properties,
+      ObjectMapper objectMapper) {
+    this.applicationDAO = applicationDAO;
+    this.properties = properties;
+    this.objectMapper = objectMapper;
+  }
+
+  @Override
+  public boolean supports(ApplicationEventListener.Type type) {
+    return properties.isEnabled() && ApplicationEventListener.Type.PRE_UPDATE == type;
+  }
+
+  @Override
+  public Application call(Application originalApplication, Application updatedApplication) {
+    boolean isChaosMonkeyEnabled = isChaosMonkeyEnabled(updatedApplication);
+    if (isChaosMonkeyEnabled(originalApplication) == isChaosMonkeyEnabled) {
+      // Flag didn't change, we don't need to do anything.
+      return updatedApplication;
+    }
+
+    Object rawPermissions = updatedApplication.details().get("permissions");
+    if (!(rawPermissions instanceof Map)) {
+      // Exit early, no permissions config found on the application
+      log.warn("No permissions config found on application '{}'", updatedApplication.getName());
+      return updatedApplication;
+    }
+
+    @SuppressWarnings("unchecked")
+    Map<String, List<String>> permissionsMap = (Map<String, List<String>>) rawPermissions;
+
+    if (isChaosMonkeyEnabled) {
```

Comment:
should exit early here if the permissionsMap is empty, think we serialize empty permissions as `{}`

Reply:
Is an empty permissions map indicative of something? Does that basically mean that everyone has access?

Reply:
Yes, it means the resource is "unrestricted".